### PR TITLE
support load go-plugin codec

### PR DESCRIPTION
**Pull request by fdingiit**
### Issues associated with this PR

Your PR should present related issues you want to solve.

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, details and the changes. At this time, Chinese is allowed to describe these.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

File: pkg/config/v2/config.go [hunk 1]
```diff
@@ -39,9 +39,10 @@ type MOSNConfig struct {
 	RawStaticResources   json.RawMessage      `json:"static_resources,omitempty"`  //static_resources raw message
 	RawAdmin             json.RawMessage      `json:"admin,omitempty"`             // admin raw message
 	Debug                PProfConfig          `json:"pprof,omitempty"`
-	Pid                  string               `json:"pid,omitempty"`     // pid file
-	Plugin               PluginConfig         `json:"plugin,omitempty"`  // plugin config
-	Extends              []ExtendConfig       `json:"extends,omitempty"` // extend config
+	Pid                  string               `json:"pid,omitempty"`              // pid file
+	Plugin               PluginConfig         `json:"plugin,omitempty"`           // plugin config
+	ThirdPartCodec       ThirdPartCodecConfig `json:"third_part_codec,omitempty"` // third part codec config
```
Comment (nejisama) [inline pkg/config/v2/config.go:44]: 可能有多个协议都使用SO方式注入

File: pkg/mosn/starter.go [hunk 2]
```diff
@@ -402,6 +406,62 @@ func initializePlugin(log string) {
...
+	fileInfo, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return err
+	}
+
+	for _, info := range fileInfo {
+		fileName := info.Name()
+		if err := readProtocolPlugin(dir + "/" + fileName); err != nil {
+			log.DefaultLogger.Warnf("[mosn] [init codec] load go plugin codec failed: %+v, err: %+v", dir+"/"+fileName, err)
+		}
```
Comment (nejisama) [inline pkg/mosn/starter.go:443]: err 和 info 应该是if else 的关系。

File: pkg/mosn/starter.go [hunk 3]
```diff
@@ -402,6 +406,62 @@ func initializePlugin(log string) {
...
+		// todo
+	default:
+		log.StartLogger.Errorf("[mosn] [init codec] unknown third part codec type: %+v", codec.Type)
+	}
+
+	log.StartLogger.Infof("[mosn] [init codec] codec all loaded")
+}
+
+const (
+	LoaderFunctionName string = "LoadCodec"
```
Comment (nejisama) [inline pkg/mosn/starter.go:430]: 这个HardCode的地方，可以考虑放到配置里

File: pkg/types/util.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package types
```
Comment (nejisama) [inline pkg/types/util.go:1]: 这个文件是哪里挪过来的？

File: pkg/config/v2/config.go (hunk 1 above)
Comment (fdingiit) [reply pkg/config/v2/config.go:44]: 是的，所以这里是扫配置里dir下所有文件，即把一个或多个codec放在同一个目录下。见initGoPluginCodec函数

File: pkg/mosn/starter.go (hunk 2 above)
Comment (fdingiit) [reply pkg/mosn/starter.go:443]: 感谢指正

File: pkg/config/v2/config.go (hunk 1 above)
Comment (fdingiit) [reply pkg/config/v2/config.go:44]: 或者我把他改成数组吧

File: pkg/mosn/starter.go (hunk 3 above)
Comment (fdingiit) [reply pkg/mosn/starter.go]: 已改

File: pkg/types/util.go (hunk 4 above)
Comment (fdingiit) [reply pkg/types/util.go:1]: constant.go拆出来的，为了不暴露StreamResetReason这个结构体，不然会有深层次的引用问题，会引用到config包等。

Comment (fdingiit) [reply pkg/types/util.go:1]: 我合回去试试

File: pkg/config/v2/config.go (hunk 1 above)
Comment (nejisama) [reply pkg/config/v2/config.go:44]: 嗯，改成数组比较合适

**Pull request closed without merging**